### PR TITLE
Enforce constant dt with dual formulation.

### PR DESCRIPTION
**Pull request by nloppi**

### CODE REVIEW

File: pyfr/integrators/dual/phys/controllers.py [hunk 1]
```diff
@@ -36,10 +37,9 @@ def advance_to(self, t):
         if t < self.tcurr:
             raise ValueError('Advance time is in the past')
 
-        while self.tcurr < t:
-            # Physical time step limiter near outputs
-            dt_lmtr = t - self.tcurr
-            dt = max(min(dt_lmtr, self._dt), self.dtmin)
+        if t - math.floor(t/self._dt)*self._dt != 0.0:
```
Comment (FreddieWitherden) [inline pyfr/integrators/dual/phys/controllers.py]: Not sure if this is stable.

File: pyfr/integrators/dual/phys/controllers.py [hunk 2]
```diff
@@ -36,10 +37,9 @@ def advance_to(self, t):
...
+            raise ValueError('End and output times must be multiplies of dt')
 
-            self.pseudointegrator.pseudo_advance(self.tcurr, dt_lmtr, dt)
-            self._accept_step(dt, self.pseudointegrator._idxcurr)
+        while self.tcurr < t:
+            self.pseudointegrator.pseudo_advance(self.tcurr, self._dt)
```
Comment (FreddieWitherden) [inline pyfr/integrators/dual/phys/controllers.py]: Should we not pass self._dt the the constructor of the pseudo integrator?

File: pyfr/integrators/dual/phys/controllers.py [hunk 3]
```diff
@@ -1,4 +1,5 @@
 # -*- coding: utf-8 -*-
+import math
```
Comment (FreddieWitherden) [inline pyfr/integrators/dual/phys/controllers.py]: Add newline.

File: pyfr/integrators/dual/pseudo/multip.py [hunk 4]
```diff
@@ -88,12 +88,12 @@ def convmon(iself, *args, **kwargs):
...
                     pass
 
-                def _rhs_with_dts(iself, t, uin, fout, c=1):
+                def _rhs_with_dts(iself, t, uin, fout):
                     # Compute -∇·f
                     iself.system.rhs(t, uin, fout)
 
                     # Coefficients for the physical stepper
-                    svals = [c*sc for sc in iself._stepper_coeffs]
+                    svals = [1/iself._dt*sc for sc in iself._stepper_coeffs]
```
Comment (FreddieWitherden) [inline pyfr/integrators/dual/pseudo/multip.py]: sc/iself._dt

File: pyfr/integrators/dual/pseudo/multip.py [hunk 5]
```diff
@@ -13,7 +13,7 @@
 
 class DualMultiPIntegrator(BaseDualPseudoIntegrator):
     def __init__(self, backend, systemcls, rallocs, mesh, initsoln, cfg,
-                 tcoeffs):
+                 tcoeffs, dt):
```
Comment (FreddieWitherden) [inline pyfr/integrators/dual/pseudo/multip.py]: Am I correct in thinking that we never use this parameter?

Comment (nloppi) [reply pyfr/integrators/dual/pseudo/multip.py]: Need to pass it onwards on line 111, to construct the pseudointegrators for each level
self.pintgs[l] = lpsint(backend, systemcls, rallocs, mesh,
                                    initsoln, cfg, tcoeffs, dt)

File: pyfr/integrators/dual/phys/controllers.py [hunk 6]
```diff
@@ -36,10 +38,10 @@ def advance_to(self, t):
         if t < self.tcurr:
             raise ValueError('Advance time is in the past')
 
-        while self.tcurr < t:
-            # Physical time step limiter near outputs
-            dt_lmtr = t - self.tcurr
-            dt = max(min(dt_lmtr, self._dt), self.dtmin)
+        rem = math.fmod(t, self._dt)
+        if rem > 1e-9 and (self._dt - rem) > 1e-9:
```
Comment (FreddieWitherden) [inline pyfr/integrators/dual/phys/controllers.py]: Do we want to use the min dt here?

**Pull request merged**